nautobot/wireless/forms.py

```diff
@@ -0,0 +1,360 @@
+from django import forms
+
+from nautobot.core.constants import CHARFIELD_MAX_LENGTH
+from nautobot.core.forms import (
+    add_blank_choice,
+    BootstrapMixin,
+    BulkEditNullBooleanSelect,
+    DynamicModelChoiceField,
+    DynamicModelMultipleChoiceField,
+    NumericArrayField,
+    StaticSelect2,
+    TagFilterField,
+)
+from nautobot.core.forms.fields import JSONArrayChoiceFormField
+from nautobot.dcim.models import Controller, Device, Location
+from nautobot.extras.forms import NautobotBulkEditForm, NautobotFilterForm, NautobotModelForm, TagsBulkEditFormMixin
+from nautobot.extras.models import SecretsGroup
+from nautobot.ipam.models import VLAN, VLANGroup
+from nautobot.tenancy.forms import TenancyFilterForm
+from nautobot.tenancy.models import Tenant
+from nautobot.wireless.models import (
+    AccessPointGroup,
+    AccessPointGroupWirelessNetworkAssignment,
+    RadioProfile,
+    SupportedDataRate,
+    WirelessNetwork,
+)
+
+from .choices import (
+    RadioProfileChannelWidthChoices,
+    RadioProfileFrequencyChoices,
+    RadioProfileRegulatoryDomainChoices,
+    SupportedDataRateStandardChoices,
+    WirelessNetworkAuthenticationChoices,
+    WirelessNetworkModeChoices,
+)
+
+
+class AccessPointGroupWirelessNetworkVLANForm(BootstrapMixin, forms.ModelForm):
+    locations = DynamicModelMultipleChoiceField(
+        queryset=Location.objects.all(),
+        required=False,
+        label="VLAN Locations (filter)",
+        null_option="None",
+    )
+    vlan_group = DynamicModelChoiceField(
+        queryset=VLANGroup.objects.all(),
+        required=False,
+        label="VLAN group (filter)",
+        null_option="None",
+        initial_params={"vlans": "$vlan"},
+    )
+    vlan = DynamicModelChoiceField(
+        queryset=VLAN.objects.all(),
+        required=False,
+        label="VLAN",
+        query_params={
+            "locations": "$locations",
+            "vlan_group": "$vlan_group",
+        },
+    )
+    access_point_group = DynamicModelChoiceField(
+        queryset=AccessPointGroup.objects.all(),
+        required=False,
+    )
+    wireless_network = DynamicModelChoiceField(
+        queryset=WirelessNetwork.objects.all(),
+        required=False,
```

Comment:
This has to do with it being used in the formset. If you are on the AccessPointGroup Form, you can only add Wireless Network and vice-versa. But if you try to mark any of the fields as required, they are required on every single line in the formset, which is definitely not what we want. It is currently possible to select a VLAN and try to save the form, but it will raise an `IntegrityError`. I spent a few hours trying to selectively mark the fields as required, but I could not get that to work. I would love some ideas though.

Reply:
Fixed, I had to implement a `clean_{field}` method on the Form. As stated above, using `required=True` impacts all fields, while the field clean method is only called if the formset form has data.